src/wallet/wallet_para_helpers.cpp

```diff
@@ -102,9 +104,147 @@ void sthd_blk_fetcher::pull_blocks(uint64_t &blocks_start_height, const std::lis
 		"mismatched blocks (" + boost::lexical_cast<std::string>(res.blocks.size()) + ") and output_indices (" +
 		boost::lexical_cast<std::string>(res.output_indices.size()) + ") sizes from daemon");
 
-	blocks_start_height = res.start_height;
-	blocks = res.blocks;
-	o_indices = res.output_indices;
+	//consider creating a temp here and assign-move
+	output.start_height = res.start_height;
+	output.blocks = std::move(res.blocks);
+	output.o_indices = std::move(res.output_indices);
+}
+
+
+mthd_blk_fetcher::mthd_blk_fetcher(size_t max_q_cap, const std::string &daemon_address, boost::optional<epee::net_utils::http::login> daemon_login) : 
+	fetcher(daemon_address, daemon_login), max_q_cap(max_q_cap), error(nullptr)
+{
+	
+}
+
+void mthd_blk_fetcher::fetch_thd_main()
+{
+	try
+	{
+		blk_batch local_data;
+		
+		while(thd_run.load(std::memory_order_relaxed)) //We will check normal exit mid-loop
+		{
+			constexpr size_t guard_hash_n = 5;
+			
+			// Prevent fetched_history from growing too large by prunning old hashes
+			// Magic number of 25 picked so that the whole req fits in an ethernet frame
+			if(fetched_history.size() > 25)
+			{
+				auto it = fetched_history.rbegin();
+				it++; //Always skip genesis hash (important)
+				
+				do
+				{ //Note we are using reverse iterators so erase will need some massaging
+					auto it_end = std::next(it, guard_hash_n);
+					it = decltype(it){ fetched_history.erase(it_end.base(), it.base()) };
+				}
+				while(fetched_history.size() > 25);
+			}
+			
+			fetcher.pull_blocks(fetched_history, local_data);
+			
+			// Daemon will always send one block which is the top of fetched_history
+			if(local_data.blocks.size() <= 1)
+				break;
+			
+			// TODO: consider checking and popping the extra block = top(fetched_history)
+			// Care - o_indicies
+			
+			// Grab up to 5 hashes from the end of the list and push them to the history list
+			// NOTE order needs to be reverse chrono
+			cryptonote::block bl;
+			size_t i = 0;
+			auto ins_pos = fetched_history.begin();
+			for (auto it = local_data.blocks.rbegin(); 
+				it != local_data.blocks.rend() && i < guard_hash_n; it++, i++)
+			{
+				bool ok = cryptonote::parse_and_validate_block_from_blob(it->block, bl);
+				THROW_WALLET_EXCEPTION_IF(!ok, error::block_parse_error, it->block);
+				
+				//Keep on inserting at the original head : 1, 2, 3, 4, 5, <head>
+				fetched_history.insert(ins_pos, cryptonote::get_block_hash(bl));
+			}
+			
+			//Lock and update mutex
+			std::unique_lock<std::mutex> lock(q_cvm);
```

Comment:
Monero code uses boost::mutex and boost::thread, not std::mutex and std::thread. std::mutex/thread are broken on Win32 and various ARM architectures and must not be used.